File: TP/CH05/CKY/python/syntaxe.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import re
import random
from typing import Dict, List, Tuple, Set


# ======================================================
#                   Analyse CKY
# ======================================================

class CKY:
    def __init__(self, gram: List[Tuple[str, str, str]], lex: Dict[str, List[str]]):
        # Grammaire
        self.gram = gram # liste((A, B, C)) où A -> B C
        self.lex = lex # mot --> liste(PoS)

    # TODO: Compléter cette méthode
    def analyser(self, phrase: List[str]) -> List[List[List[Tuple[str, int, int, int]]]]:
        T = []
        N = len(phrase)
        for i in range(N):
            T.append([[] for i in range(N)])
            mot = phrase[i]
            for A in self.lex[mot]: # ici, on suppose que le mot existe dans la lexique ; sinon, on aura une erreur
                T[i][i].append((A, -1, -1, -1)) # On ajoute toutes les variables possibles
            
        # Compléter ici : remplissage 

        return T

    def exporter_json(self):
        return self.__dict__.copy()

    def importer_json(self, data):
        for cle in data:
            self.__dict__[cle] = data[cle]

# ======================================================
#                Analyse syntaxique
# ======================================================

# TODO: compléter cette fonction
def evaluer_stricte(ref, sys):
    return 0

def construire(T, phrase, i, j, pos):
    A, k, iB, iC = T[i][j][pos]
    #A = A.upper()
    if k >= 0:
        gauche = construire(T, phrase, i, k, iB)
        droit = construire(T, phrase, k+1, j, iC)
        return (A, gauche, droit)
    return (A, phrase[i])

def parse_tuple(string):
    string = re.sub('([^\s(),]+)', "'\\1'", string)
    #print(string)
    try:
        s = eval(string)
        if type(s) == tuple:
            return s
        return
    except:
        return

class Syntaxe():
    def __init__(self):
        self.eval = []

    def _analyser(self, phrase):
        r = None
        T = self.modele.analyser(phrase)
        n = len(phrase) - 1
        # for i in range(n+1):
        #     for j in range(i, n+1):
        #         print(i+1, j+1, T[i][j])
        for pos in range(len(T[0][n])):
            if T[0][n][pos][0] == 'S':
                # print('bingo')
                r = construire(T, phrase, 0, n, pos)
                break
        return r

    def analyser(self, phrase: str):
        return self._analyser(phrase.strip().lower().split())

    def charger_modele(self, url):
        f = open(url, 'r')
        lex = {}
        gram = []
        for l in f: # la lecture ligne par ligne
            l = l.strip()
            if len(l) < 5 or l.startswith('#') :
                continue
            info = l.split('	')
            if len(info) == 2:
                if not info[1] in lex:
                    lex[info[1]] = []
                lex[info[1]].append(info[0])
            elif len(info) == 3:
                gram.append((info[0], info[1], info[2]))
        self.modele = CKY(gram, lex)
        f.close()

    def charger_evaluation(self, url):
        f = open(url, 'r')
        for l in f: # la lecture ligne par ligne
            l = l.strip()
            if len(l) < 5 or l.startswith('#'):
                continue
            info = l.split('	')
            
            self.eval.append((info[0], parse_tuple(info[1])))

    def evaluer(self, n):
        if n == -1:
            S = self.eval
            n = len(S)
        else :
            S = random.sample(self.eval, n)
        score = 0.0
        for i in range(n):
            test = S[i]
            print('=======================')
            print('phrase :', test[0])
            print('arbre de référence :', test[1])
            arbre = self.analyser(test[0])
            print('arbre du système :', arbre)
            score_i = evaluer_stricte(test[1], arbre)
            print('score :', score_i)
            score += score_i

        score = score/n
        print('---------------------------------')
        print('score total : ', score)



# ======================================================
#        Génération du graphique (langage Dot)
# ======================================================

def generer_noeud(noeud, id=0):
    # Si le noeud n'existe pas
    if noeud is None:
        return 0, ''
    # Si le noeud est final,
    nid = id + 1
    if len(noeud) < 3:
        return nid, 'N' + str(id) + '[label="' + noeud[0] + "=" + noeud[1] + '" shape=box];\n'
    # Sinon,
    # s'il y a des fils, on boucle sur les fils et on imprime des SI ... ALORS
    res = 'N' + str(id) + '[label="' + noeud[0] + '"];\n'
    nid_g = nid
    nid, code = generer_noeud(noeud[1], id=nid_g)
    res += code
    res += 'N' + str(id) + ' ->  N' + str(nid_g) + ';\n'
    nid_d = nid
    nid, code = generer_noeud(noeud[2], id=nid_d)
    res += code
    res += 'N' + str(id) + ' ->  N' + str(nid_d) + ';\n'
    return nid, res

def generer_graphviz(racine, url):
    res = 'digraph Tree {\n'
    id, code = generer_noeud(racine)
    res += code
    res += '}'
    f = open(url, 'w')
    f.write(res)
    f.close()


# ======================================================
#                       TESTES
# ======================================================

# le test du cours pour tester l'analyseur CKY
def tester_cky():
    parser = Syntaxe()
    parser.charger_modele('data/gram1.txt')
    phrase = 'la petite forme une petite phrase'
    resultat = "('S', ('NP', ('DET', 'la'), ('N', 'petite')), ('VP', ('V', 'forme'), ('NP', ('DET', 'une'), ('AP', ('ADJ', 'petite'), ('N', 'phrase')))))"
    arbre = parser.analyser(phrase)
    print('résultat attendu : ', resultat)
    print('mon     résultat : ', arbre)
    generer_graphviz(arbre, 'arbre_python.gv')

# tester l'évaluation stricte
def tester_evaluer_arbre():
    print('résultat attendu : 0, résultat trouvé : ', evaluer_stricte(None, ('A', 'd')))
    print('résultat attendu : 1, résultat trouvé : ', evaluer_stricte(None, None))
    print('résultat attendu : 0, résultat trouvé : ', evaluer_stricte(('A', 'a'), ('A', ('B', 'b'), ('C', 'c'))))
    print('résultat attendu : 0, résultat trouvé : ', evaluer_stricte(('A', ('B', 'b'), ('C', 'c')), ('A', 'a')))
    print('résultat attendu : 0, résultat trouvé : ', evaluer_stricte(('A', ('B', 'b'), ('C', 'c')), ('A', ('B', 'd'), ('C', 'c'))))
    print('résultat attendu : 1, résultat trouvé : ', evaluer_stricte(('A', ('B', 'b'), ('C', 'c')), ('A', ('B', 'b'), ('C', 'c'))))

# tester l'évaluation stricte
# Le résultat doit être 1
def tester_evaluation():
    parser = Syntaxe()
    parser.charger_modele('data/gram1.txt')
    parser.charger_evaluation('data/test1.txt')
    parser.evaluer(-1)


if __name__ == '__main__':
    tester_cky()
    #tester_evaluer_arbre()
    #tester_evaluation()

File: TP/CH05/CKY/java/Syntaxe.java
import java.util.regex.*;
import java.io.*;
import java.util.*;
import java.lang.Math;
import java.util.stream.*;

public class Syntaxe {
	private CKY modele;
	private List<Tuple<String, Noeud>> eval = new ArrayList<>();

	public Noeud analyser(String[] phrase){
		List<List<List<CKYRef>>> T = this.modele.analyser(phrase);
		int N = phrase.length - 1;
		List<CKYRef> lastCell = T.get(0).get(N);
		for(int pos=0; pos<lastCell.size(); pos++){
			if (lastCell.get(pos).A.equals("S")){
				return Outils.construire(T, phrase, 0, N, pos);
			}
		}
		return null;
	}

	public Noeud analyser(String phrase){
		return analyser(phrase.split(" "));
	}

	public void charger_modele(String url){
		try {
			BufferedReader f = new BufferedReader(new FileReader(url));
			String l="";
			Map<String, List<String>> lex = new HashMap<>();
			List<Regle> gram = new ArrayList<>();
			while((l = f.readLine()) != null) {
				l = l.strip();
				if ((l.length() < 5)) continue;
				String[] info = l.split("	");
				if (info.length == 2){
					if (! lex.containsKey(info[1])) lex.put(info[1], new ArrayList<String>());
					lex.get(info[1]).add(info[0]);
				} else if (info.length==3){
					gram.add(new Regle(info[0], info[1], info[2]));
				}
			}
			f.close();
			this.modele = new CKY(gram, lex);
		} catch(IOException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}


	public void charger_evaluation(String url){
		try {
			BufferedReader f = new BufferedReader(new FileReader(url));
			String l;
			while((l = f.readLine()) != null) {
				l = l.strip();
				if ((l.length() < 5)) continue;
				String[] info = l.split("	");
				this.eval.add(new Tuple<String, Noeud>(info[0], Outils.parse_tuple(info[1])));
			}
			f.close();
		} catch(IOException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	public void evaluer(int n){
		if (n==-1) n = this.eval.size();
		else Collections.shuffle(this.eval);

		double score = 0.0;
		for (int i=0; i<n; i++){
			Tuple<String, Noeud> test = this.eval.get(i);
			System.out.println("=======================");
			System.out.println("phrase : " + test.k);
			System.out.println("arbre de référence : " + test.v);
			Noeud arbre = analyser(test.k);
			System.out.println("arbre du système   : " + arbre);
			int score_i = Outils.evaluer_stricte(test.v, arbre);
            System.out.println("score :" + score_i);
            score += score_i;
		}

		score /= n;
        System.out.println("---------------------------------");
        System.out.println("score total : " + score);
  }


	// ======================================================
	//                       TESTES
	// ======================================================

	public static void tester_cky(){
		Syntaxe parser = new Syntaxe();
		parser.charger_modele("data/gram1.txt");
		String phrase = "la petite forme une petite phrase";
		Noeud arbre = parser.analyser(phrase);
		String resultat = "('S', ('NP', ('DET', 'la'), ('N', 'petite')), ('VP', ('V', 'forme'), ('NP', ('DET', 'une'), ('AP', ('ADJ', 'petite'), ('N', 'phrase')))))";
    	arbre = parser.analyser(phrase);
    	System.out.println("résultat attendu : " + resultat);
    	System.out.println("mon     résultat : " + arbre);
		generer_graphviz(arbre, "arbre_java.gv");
	}

	public static void tester_evaluer_arbre(){
		List<Integer> resultat = new ArrayList<>();
		List<Noeud> test1 = new ArrayList<>();
		List<Noeud> test2 = new ArrayList<>();

		test1.add(null);
		test2.add(new Noeud("A", new Noeud("d", null, null), null));
		resultat.add(0);

		test1.add(null);
		test2.add(null);
		resultat.add(1);

		test1.add(new Noeud("A", new Noeud("a", null, null), null));
		test2.add(new Noeud("A", new Noeud("B", new Noeud("b", null, null), null), new Noeud("C", new Noeud("c", null, null), null)));
		resultat.add(0);

		test1.add(new Noeud("A", new Noeud("B", new Noeud("b", null, null), null), new Noeud("C", new Noeud("c", null, null), null)));
		test2.add(new Noeud("A", new Noeud("a", null, null), null));
		resultat.add(0);

		test1.add(new Noeud("A", new Noeud("B", new Noeud("b", null, null), null), new Noeud("C", new Noeud("c", null, null), null)));
		test2.add(new Noeud("A", new Noeud("B", new Noeud("d", null, null), null), new Noeud("C", new Noeud("c", null, null), null)));
		resultat.add(0);

		test1.add(new Noeud("A", new Noeud("B", new Noeud("b", null, null), null), new Noeud("C", new Noeud("c", null, null), null)));
		test2.add(new Noeud("A", new Noeud("B", new Noeud("b", null, null), null), new Noeud("C", new Noeud("c", null, null), null)));
		resultat.add(1);

		for (int i=0; i<resultat.size(); i++){
			System.out.print("résultat attendu : " + resultat.get(i));
			System.out.println(", résultat trouvé : " + evaluer_stricte(test1.get(i), test2.get(i)));
		}

	}

	public static void tester_evaluation(){
		Syntaxe parser = new Syntaxe();
		parser.charger_modele("data/gram1.txt");
		parser.charger_evaluation("data/test1.txt");
		parser.evaluer(-1);
	}

	public static void main(String[] args) {
		//tester_cky();
		//tester_evaluer_arbre();
		tester_evaluation();
	}
}

class CKY {
	private List<Regle> gram;
	private Map<String, List<String>> lex;

	public CKY(List<Regle> gram, Map<String, List<String>> lex){
		this.gram = gram;
		this.lex = lex;
	}

	//TODO: Compléter cette méthode
	public List<List<List<CKYRef>>> analyser(String[] phrase){
		int N = phrase.length;
		List<List<List<CKYRef>>> T = new ArrayList<>();
		for(int i=0; i<N; i++){
			List<List<CKYRef>> Ti = new ArrayList<>();
			T.add(Ti);
			for(int j=0; j<N; j++) Ti.add(new ArrayList<CKYRef>());
			String mot = phrase[i];
			for (String A : this.lex.get(mot)){
				T.get(i).get(i).add(new CKYRef(A, -1, -1, -1));
			}
		}

		// Compléter ici : remplissage 
		return T;
	}

}

class Outils {

	//TODO: compléter cette fonction
	public static int evaluer_stricte(Noeud ref, Noeud sys){
    	return 0;
	}

	public static Noeud construire(List<List<List<CKYRef>>> T, String[] phrase, int i, int j, int pos){
		CKYRef ref = T.get(i).get(j).get(pos);
		String A = ref.A.toUpperCase();
		if (ref.k >= 0){
			Noeud gauche = construire(T, phrase, i, ref.k, ref.iB);
			Noeud droit = construire(T, phrase, ref.k + 1, j, ref.iC);
			return new Noeud(A, gauche, droit);
		}
		return new Noeud(A, new Noeud(phrase[i], null, null), null);
	}

	public static String generer_noeud(Noeud noeud){
		int id = -1;
		int pid = 0;
		Deque<NoeudPere> pile = new ArrayDeque<>();
		Noeud n = noeud;
		String res = "";
		while(pile.size() > 0 || n != null){
			id += 1;
			if (n==null){
				NoeudPere npid = pile.pop();
				n = npid.n;
				pid = npid.pid;
			} else if (n.gauche == null){
				res += "N" + id +"[label=\"" + n.val + "\" shape=box];\n";
				if (pid < id) res += "N" + pid + " ->  N" + id + ";\n";
				n = null;
			} else {
				res += "N" + id + "[label=\"" + n.val + "\"];\n";
				if (pid < id) res += "N" + pid + " ->  N" + id + ";\n";
				//ce noeud sera un parent
				pid = id;
				//empiler la droite
				pile.push(new NoeudPere(n.droit, pid));
				//aller toujours vers la gauche
				n = n.gauche;
			}
		}
		return res;
	}

	public static void generer_graphviz(Noeud racine, String url){
		String res = "digraph Tree {\n";
		res += generer_noeud(racine);
		res += "}";
		try {
			BufferedWriter f = new BufferedWriter(new FileWriter(url));
			f.write(res);
			f.close();
		} catch (IOException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	private static int trouver_par_ferme(String string){
		LinkedList<Character> pile = new LinkedList<>();
		for(int i = 0;i<string.length();i++){
			char c = string.charAt(i);
			if(c == '(') pile.push(c);
			else if (c == ')') pile.pop();

			if(pile.size() == 0) return i;
		}
		return -1;
	}

	public static Noeud parse_tuple(String string){
		string = string.replace(" ", "");
		if (! string.startsWith("(")) return null;
		string = string.substring(1, string.length()-1);
		int idxVirgule = string.indexOf(",");
		if (idxVirgule == -1) return null;
		String val = string.substring(0, idxVirgule);
		string = string.substring(idxVirgule+1);
		if (! string.startsWith("(")) return new Noeud(val, new Noeud(string, null, null), null);
		
		
		int idxParFerm = trouver_par_ferme(string);
		if (idxParFerm == -1) return null;
		Noeud gauche = parse_tuple(string.substring(0, idxParFerm+1));
		Noeud droit = null;
		if (idxParFerm < string.length()) droit = parse_tuple(string.substring(idxParFerm+2));

		return new Noeud(val, gauche, droit);
	}

}

class CKYRef {
	public final String A;
	public final int k;
	public final int iB;
	public final int iC;
	public CKYRef(String A, int k, int iB, int iC){
		this.A = A;
		this.k = k;
		this.iB = iB;
		this.iC = iC;
	}

	public String toString(){
		return "(" + A + ", " + k + ", " + iB + ", " + iC + ")";
	}
}

class Regle {
	public final String A;
	public final String B;
	public final String C;
	public Regle(String a, String b, String c) {
		this.A = a;
		this.B = b;
		this.C = c;
	}

	public String toString(){
		return A + " -> " + B + " " + C;
	}
}

class Noeud {
	public final String val;
	public final Noeud gauche;
	public final Noeud droit;
	public Noeud(String val, Noeud gauche, Noeud droit) {
		this.val = val;
		this.gauche = gauche;
		this.droit = droit;
	}

	public int len(){
		if (gauche != null && droit != null) return 3;
		if (gauche != null) return 2;
		if (gauche == null && droit == null) return 1;
		return 0;
	}

	public String toString(){
		if (gauche == null && droit == null) return "'" + val + "'";

		String res = "('" + val + "'";
		res += ", " + gauche;
		if (droit != null) res += ", " + droit;
		res += ")";
		return res;
	}
}

class NoeudPere {
	public final Noeud n;
	public final int pid;
	public NoeudPere(Noeud n, int pid) {
		this.n = n;
		this.pid = pid;
	}
}

class Tuple<K, V> {
	public final K k;
	public final V v;
	public Tuple(K k, V v) {
		this.k = k;
		this.v = v;
	}

	public String toString(){
		return "(" + k + ", " + v + ")";
	}
}